DOCS-3423: Update training script to parse labels based on feedback from etai/tahiya (#3941)
viamrobotics/docs: docs/data-ai/ai/train.md

<p><strong>4. Add <code>training.py</code> code</strong></p>

> <p>Copy one of the following templates into <file>training.py</file>, depending on whether or not you wish to parse labels:</p>

Reviewer: Can you rephrase this to have the condition based on which the reader should decide be clearer? Why would a user need to use the labels flag or not?

> <p>Copy one of the following templates into <file>training.py</file>, depending on whether or not you wish to parse labels:</p>

Author: ```suggestion
<p>Copy one of the following templates into <file>training.py</file>, depending on whether or not you wish to parse labels instead of having to hard code the labels in the training script and re-upload for each set of labels, allowing you to more easily reuse training scripts:</p>
```

> <p>Copy one of the following templates into <file>training.py</file>, depending on whether or not you wish to parse labels instead of having to hard code the labels in the training script and re-upload for each set of labels, allowing you to more easily reuse training scripts:</p>

Author: ```suggestion
<p>Copy one of the following templates into <file>training.py</file>, depending on whether or not you wish to parse labels instead of having to hard code the labels in the training script and re-upload for each set of labels, which allows you to more easily reuse training scripts:</p>
```

> <p>Copy one of the following templates into <file>training.py</file>, depending on whether or not you wish to parse labels instead of having to hard code the labels in the training script and re-upload for each set of labels, which allows you to more easily reuse training scripts:</p>

Reviewer: This is now a very long sentence and it makes it hard to understand where labels come from.

```suggestion
<p>You can set up your training script to use a hard coded set of labels or allow users to pass in a set of labels when using the training script. Allowing users to pass in labels when using training scripts makes your training script more flexible for reuse.</p>
<p>Copy one of the following templates into <file>training.py</file>, depending on how you want to handle labels:</p>
```

Do you see the difference?

Author: Yes, I do see the difference. Will keep that in mind. Thank you!

API_KEY = os.environ['API_KEY']
API_KEY_ID = os.environ['API_KEY_ID']

DEFAULT_EPOCHS = 200

# This parses the required args for the training script.
# The model_dir variable will contain the output directory where
# the ML model that this script creates should be stored.
# The data_json variable will contain the metadata for the dataset
# that you should use to train the model.

> 

Author: know this is awk but python markdown parser was complaining about this specifically

    if present. These must be parsed as command line arguments and then used
    as the model input and output, respectively. The number of epochs can be
    used to optionally override the default.
    """
    parser = argparse.ArgumentParser()
    parser.add_argument("--dataset_file", dest="data_json",
                        type=str, required=True)
    parser.add_argument("--model_output_directory", dest="model_dir",
                        type=str, required=True)
    parser.add_argument("--num_epochs", dest="num_epochs", type=int)
    parser.add_argument(
        "--labels",
        dest="labels",
        type=str,
        required=True,
        help="Space-separated list of labels, \
            enclosed in single quotes (e.g., 'label1 label2').",
    )
    args = parser.parse_args()

    if not args.labels:
        raise ValueError("Labels must be provided")

    labels = [label.strip() for label in args.labels.strip("'").split()]
    return args.data_json, args.model_dir, args.num_epochs, labels

# This is used for parsing the dataset file (produced and stored in Viam),
# parse it to get the label annotations
# Used for training classifiction models

> 

Author: flake8 parser was complaining about this

# Build the model
def build_and_compile_model(

>     labels: ty.List[str], model_type: str, input_shape: ty.Tuple[int, int, int]) -> Model:

Author: ```suggestion
    labels: ty.List[str], model_type: str, input_shape: ty.Tuple[int, int, int]
) -> Model:
```

        model_dir: output directory for model artifacts
        model_name: name of saved model
    """

>     file_type = "tflite"  # Add proper file type

Author: ```suggestion
    # Save the model to the output directory
    file_type = "tflite"  # Add proper file type
```

    # Save the model to the output directory
    file_type = "tflite"  # Add proper file type
    filename = os.path.join(model_dir, f"{model_name}.{file_type}")

    # Example: Convert to TFLite
    converter = tf.lite.TFLiteConverter.from_keras_model(model)
    tflite_model = converter.convert()

    # Save the model
    with open(filename, "wb") as f:
        f.write(tflite_model)

if __name__ == "__main__":
    DATA_JSON, MODEL_DIR, NUM_EPOCHS, LABELS = parse_args()

    IMG_SIZE = (256, 256)

    # Read dataset file.
    # The model type can be changed based on whether you want the model to
    # output one label per image or multiple labels per image
    model_type = multi_label
    image_filenames, image_labels = parse_filenames_and_labels_from_json(
        DATA_JSON, LABELS, model_type)

    # Validate epochs
    epochs = (
        DEFAULT_EPOCHS if NUM_EPOCHS is None
        or NUM_EPOCHS <= 0 else int(NUM_EPOCHS)
    )

    # Build and compile model on data
    model = build_and_compile_model(image_labels, model_type, IMG_SIZE + (3,))

    # Save labels.txt file
    save_labels(LABELS + [unknown_label], MODEL_DIR)
    # Convert the model to tflite
    save_model(

>         model, MODEL_DIR, "classification_model"

Author: correction, was never being used

    image_filenames, image_labels = parse_filenames_and_labels_from_json(
        DATA_JSON, LABELS, model_type)

    # Validate epochs
    epochs = (
        DEFAULT_EPOCHS if NUM_EPOCHS is None
        or NUM_EPOCHS <= 0 else int(NUM_EPOCHS)
    )

    # Build and compile model on data
    model = build_and_compile_model(image_labels, model_type, IMG_SIZE + (3,))

    # Save labels.txt file
    save_labels(LABELS + [unknown_label], MODEL_DIR)
    # Convert the model to tflite
    save_model(
        model, MODEL_DIR, "classification_model"
    )
```

{{% /expand %}}

> {{% expand "Click to see the template with labels parsed" %}}

Author: ```suggestion
{{% expand "Click to see the template with parsed labels" %}}
```

- `dataset_file`: a file containing the data and metadata for the training job
- `model_output_directory`: the location where the produced model artifacts are saved to

> If you used the version with labels parsed, it will also take the following command line inputs:

Reviewer: ```suggestion
If you used the version with parsed labels, it will also take the following command line inputs:
```

> If you used the version with parsed labels, it will also take the following command line inputs:

Reviewer: ```suggestion
If you used the training script template that allows users to pass in lables, it will also take the following command line inputs:
```

Click **Train model** and select **Train on a custom training script**, then follow the prompts.

{{% alert title="Tip" color="tip" %}}

> If you used the version of <file>training.py</file> with labels parsed, your training job will fail with the error `ERROR training.py: error: the following arguments are required: --labels`.

Reviewer: ```suggestion
If you used the version of <file>training.py</file> with parsed labels, your training job will fail with the error `ERROR training.py: error: the following arguments are required: --labels`.
```

> If you used the version of <file>training.py</file> with parsed labels, your training job will fail with the error `ERROR training.py: error: the following arguments are required: --labels`.

Reviewer: ```suggestion
If you used the version of <file>training.py</file> that allows users to pass in lables, your training job will fail with the error `ERROR training.py: error: the following arguments are required: --labels`.
```

> Use the CLI and enter labels.

Reviewer: ```suggestion
To use labels, you must use the CLI.
```
